numba/np/ufunc/dufunc.py

```diff
@@ -778,6 +779,136 @@ def impl_axis_none(ufunc,
+                                          axis,
+                                          dtype=dtype,
+                                          out=out)
+                return impl_cast
+
+            @register_jitable
+            def tuple_slice(tup, pos):
+                s = tup_m1  # TODO: replace this by a literally call
+                i, j = 0, 0
+                while i < len(tup):
+                    if i != pos:
+                        s = tuple_setitem(s, j, tup[i])
+                        j += 1
+                    i += 1
+                return s
+
+            # importing it here as an import at the top scope brings unwanted
+            # stuff. See numba/tests/test_import.py::test_no_impl_import
+            from numba.np.arrayobj import generate_getitem_setitem_with_axis
+            setitem = generate_getitem_setitem_with_axis(array.ndim, 'setitem')
+
+            def impl(ufunc, array, indices, axis=0, dtype=None, out=None):
+                sz = indices.shape[0]
+
+                ax = axis
+                if axis < 0:
+                    axis += array_ndim
+
+                if axis < 0 or axis >= array_ndim:
+                    raise ValueError(f"axis {ax} is out of bounds for array "
+                                     f"of dimension {array_ndim}")
+
+                shape = tuple_setitem(array.shape, axis, sz)
+
+                if not out_none and out.shape != shape:
+                    # TODO: improve error message
+                    msg = ('operands could not be broadcast together with '
+                           'remmaped shapes')
+                    raise ValueError(msg)
+
+                if out_none:
+                    out = np.zeros(shape, dtype=dt)
+
+                # short-circuit to avoid overflow on Windows
+                if len(indices) == 0:
+                    return out
+                # print(indices)
+
+                j = 0
+                for i in range(len(indices) - 1):
+                    if indices[i] < indices[i + 1]:
+                        idx = np.arange(indices[i], indices[i + 1])
+                        if array_ndim > 1:
+                            arr_slice = np.take(array, idx, axis)
+                        else:
+                            arr_slice = array[idx]
+                        arr_reduce = ufunc.reduce(arr_slice, axis=axis)
+                        setitem(out, j, axis, arr_reduce)
+                    elif indices[i] >= indices[i + 1]:
+                        idx = indices[i]
+                        arr_slice = np.take(array, idx, axis)
+                        if array_ndim > 1:
+                            _slice = tuple_slice(array.shape, axis)
+                            arr_slice = arr_slice.reshape(_slice)
+                        setitem(out, j, axis, arr_slice)
+                    elif indices[i] >= sz or indices[i] < 0:
+                        # TODO: Improve error message
+                        raise ValueError('error!')
```

Comment:
The error message here needs to be improved before we can merge.

Reply:
Done!